virttest/test_setup.py

```diff
@@ -97,6 +100,90 @@ class PolkitConfigCleanupError(PolkitConfigError):
     pass
 
 
+class Setuper(object):
+
+    """
+    Virtual base abstraction of setuper.
+    """
+
+    __metaclass__ = ABCMeta
+
+    def __init__(self, test, params, env):
+        """
+        Initializes the setuper.
+
+        :param test: VirtTest instance.
+        :param params: Dictionary with the test parameters.
+        :param env: Dictionary with test environment.
+        """
+        self.test = test
+        self.params = params
+        self.env = env
+
+    @abstractmethod
+    def setup(self):
+        """Setup procedure."""
+        raise NotImplementedError
+
+    @abstractmethod
+    def cleanup(self):
+        """Cleanup procedure."""
+        raise NotImplementedError
+
+
+class SetuperMgr(object):
+
+    """
+    Setuper Manager implementation.
+
+    The instance can help do the setup stuff before test started and
+    do the cleanup stuff after test finished. This setup-cleanup
+    combined stuff will be performed in LIFO order.
+    """
+
+    def __init__(self, test, params, env):
+        """
+        Initializes the setuper manager.
+
+        :param test: VirtTest instance.
+        :param params: Dictionary with the test parameters.
+        :param env: Dictionary with test environment.
+        """
+        self._stp_args = (test, params, env)
+        self._reg_stps = []
+        self._setupers = []
```

Comment:
Oh, `setupers` are here. Do you expect people using `do_setup` multiple times? If not then on register you can do the `init` and store only `_setupers`.

Reply:
I think it'd actually be better. The `register` would `init` the objects and store in `_setupers`. Then `do_setup` would go one by one and run `setup()` on them and would __always__ run the cleanup on all registered items (not just executed ones; plugins would be free to skip it in case setup was not executed).

Reply:
Previously, I did make the approach which your suggested, then I realized that it is better to execute `init` and `setup` together for each setuper because some of the `init` is not just a init, and one setuper may depend on another setuper. I am not sure whether we should accept that kind of implementation (or maybe have to?), so I chose the current approach.

Reply:
Ideally I'd require all `__init__` to be just simple ones without any additional steps, but alternative solution where `__init__` is called first time it's used (be it `setup` or `cleanup`) works for me as long as they are always executed. That is the important bit... (even though in most cases we won't use that feature, some setupers will need it)

Reply:
(there could actually be second plugin type to have it executed only when setup was executed to avoid unnecessary boilerplate code...)